## **Случайные события. Практическая работа**

### **Задание 1**

Даны таблицы с суммами и произведениями значений при броске двух кубиков. Используя эти таблицы, вычислите вероятности следующих сложных событий.

<p align="left">
  <img src="https://drive.google.com/uc?id=1If1NDfwPCfLC6JHqxa5IDgXLRzrbB-sI" alt="drawing" width="600"/>
</p>

<p align="left">
  <img src="https://drive.google.com/uc?id=1vWNTqND6br6XzESVSwqKjK9AqeN0-k-4" alt="drawing" width="600"/>
</p>

1. Вероятность, что полученная сумма значений на кубиках — нечётное число.


In [1]:
n = 36 #размер таблицы, мощность пространства элементарных событий 

# выберем по таблице из всего множества событий те, где "в сумме поулчили нечетное число"
A_odd = {(1,2),(1,4),(1,6),(2,1),(2,3),(2,5),(3,2),(3,4),(3,6),(4,1),(4,3),(4,5),(5,2),(5,4),(5,6),(6,1),(6,3),(6,5)}

print("Число исходов =", len(A_odd), "; Вероятность получить в сумме нечетное число =", len(A_odd) / n)

Число исходов = 18 ; Вероятность получить в сумме нечетное число = 0.5


2. Вероятность, что полученная сумма значений на кубиках — число, кратное трём.



In [2]:
# полный список элементарных исходов из первой таблицы
A = [(i,j,i+j) for i in range(1,7) for j in range(1,7)]

# выберем элементарные исходы про "сумма значений на кубиках — число, кратное трём"
A_mult3 = {elem for elem in A if not elem[2] % 3}
print("Множество исходов, где сумма значений на кубиках — число, кратное трём\n", A_mult3)

print("\nЧисло исходов =", len(A_mult3), "; Вероятность получить число, кратное трём =", round(len(A_mult3) / n, 3))

Множество исходов, где сумма значений на кубиках — число, кратное трём
 {(5, 1, 6), (6, 6, 12), (6, 3, 9), (1, 2, 3), (2, 1, 3), (3, 6, 9), (1, 5, 6), (3, 3, 6), (4, 5, 9), (4, 2, 6), (2, 4, 6), (5, 4, 9)}

Число исходов = 12 ; Вероятность получить число, кратное трём = 0.333


3. Вероятность, что сумма будет меньше пяти или больше десяти.


In [3]:
# выберем элементарные исходы про "сумма меньше 5"
A_less5 = {elem for elem in A if elem[2] < 5}
print("Множество исходов с суммой меньше 5\n", A_less5)

# выберем элементарные исходы про "сумма больше 10"
A_greater10 = {elem for elem in A if elem[2] > 10}
print("\nМножество исходов с суммой больше 10\n", A_greater10)

# убедимся, что события незвисимы и можно будет просто сложить вероятности
print("\nУсловия независимы:", len(A_less5 & A_greater10) == 0)

# значит, общая вероятность получается обычным сложением
A_task3 = A_less5 | A_greater10

print("\nЧисло исходов =", len(A_task3), "; Вероятность получить описанное число =", round(len(A_task3) / n, 3))

Множество исходов с суммой меньше 5
 {(3, 1, 4), (1, 2, 3), (2, 1, 3), (1, 3, 4), (1, 1, 2), (2, 2, 4)}

Множество исходов с суммой больше 10
 {(6, 5, 11), (6, 6, 12), (5, 6, 11)}

Условия независимы: True

Число исходов = 9 ; Вероятность получить описанное число = 0.25


4. Вероятность, что произведение значений будет равно двенадцати.


In [4]:
# полный список элементарных исходов из второй таблицы
B = [(i,j,i*j) for i in range(1,7) for j in range(1,7)]

# выберем элементарные исходы про "произведение значений будет равно двенадцати"
B_mult12 = {elem for elem in B if elem[2] == 12}
print("Множество исходов, где произведение значений будет равно двенадцати:\n", B_mult12)

print("\nЧисло исходов =", len(B_mult12), "; Вероятность получить описанное число =", round(len(B_mult12) / n, 3))

Множество исходов, где произведение значений будет равно двенадцати:
 {(2, 6, 12), (4, 3, 12), (3, 4, 12), (6, 2, 12)}

Число исходов = 4 ; Вероятность получить описанное число = 0.111


5. Вероятность, что произведение значений — нечётно.


In [5]:
# выберем элементарные исходы про "произведение значений — нечётно"
B_odd = {elem for elem in B if elem[2] % 2}
print("Множество исходов, где произведение значений — нечётно:\n", B_odd)

print("\nЧисло исходов =", len(B_odd), "; Вероятность получить описанное число =", round(len(B_odd) / n, 3))

Множество исходов, где произведение значений — нечётно:
 {(5, 1, 5), (3, 1, 3), (5, 3, 15), (3, 5, 15), (3, 3, 9), (1, 3, 3), (1, 1, 1), (1, 5, 5), (5, 5, 25)}

Число исходов = 9 ; Вероятность получить описанное число = 0.25


6. Вероятность, что сумма значений окажется больше или равной их произведению.

In [7]:
# выберем элементарные исходы про "сумма значений >= их произведения"
AgreaterB = {elemA for elemA,elemB in zip(A,B) if elemA[2] >= elemB[2]}
print("Множество исходов, где сумма значений >= их произведения:\n", AgreaterB)

print("\nЧисло исходов =", len(AgreaterB), "; Вероятность получить такой исход =", round(len(AgreaterB) / n, 3))

Множество исходов, где сумма значений >= их произведения:
 {(5, 1, 6), (3, 1, 4), (4, 1, 5), (1, 2, 3), (2, 1, 3), (1, 3, 4), (1, 1, 2), (1, 5, 6), (2, 2, 4), (6, 1, 7), (1, 4, 5), (1, 6, 7)}

Число исходов = 12 ; Вероятность получить такой исход = 0.333


### **Задание 2**
Загрузите данные о пассажирах «Титаника» из файла titanic.csv в приложенных к практической работе файлах. Ограничьтесь набором таких переменных, как класс билета пассажира, пол и переменной, показывающей, выжил он или нет. В этом датасете все эти переменные заполнены непустым значением.


In [8]:
import pandas as pd
titanic = pd.read_csv("train.csv")
X = titanic[['PassengerId','Survived','Pclass','Sex']]
X

PassengerId  Survived  Pclass     Sex
0              1         0       3    male
1              2         1       1  female
2              3         1       3  female
3              4         1       1  female
4              5         0       3    male
..           ...       ...     ...     ...
886          887         0       2    male
887          888         1       1  female
888          889         0       3  female
889          890         1       1    male
890          891         0       3    male

[891 rows x 4 columns]

1. Запишите в P_fem вероятность, что случайный пассажир «Титаника» — женщина.


In [9]:
sex_data = X["Sex"].value_counts()
P_fem = sex_data["female"] / sum(sex_data)

print(sex_data, "\nВероятность, что случайный пассажир «Титаника» - женщина =", round(P_fem, 3))

Sex
male      577
female    314
Name: count, dtype: int64 
Вероятность, что случайный пассажир «Титаника» - женщина = 0.352


2. Запишите в P_1class вероятность, что случайный пассажир «Титаника» ехал в первом классе.


In [10]:
class_data = X["Pclass"].value_counts()
P_1class = class_data[1] / sum(class_data)

print(class_data, "\nВероятность, что случайный пассажир «Титаника» ехал в первом классе =", round(P_1class, 3))

Pclass
3    491
1    216
2    184
Name: count, dtype: int64 
Вероятность, что случайный пассажир «Титаника» ехал в первом классе = 0.242


3. Запишите в P_surv вероятность, что случайный пассажир «Титаника» выжил.


In [11]:
surv_data = X["Survived"].value_counts()
P_surv = surv_data[1] / sum(surv_data)

print(surv_data, "\nВероятность, что случайный пассажир «Титаника» выжил =", round(P_surv, 3))

Survived
0    549
1    342
Name: count, dtype: int64 
Вероятность, что случайный пассажир «Титаника» выжил = 0.384


4. Запишите в P_fem_surv вероятность, что пассажир — женщина, и одновременно с этим пассажир выжила в катастрофе.


In [12]:
X_t4 = X[(X["Survived"] == 1) & (X["Sex"] == "female")]
print("Подмножество исходов про пассажир — женщина, и одновременно с этим пассажир выжила в катастрофе:\n", X_t4)

P_fem_surv = X_t4.shape[0] / X.shape[0]
print("\nВероятность такого исхода =", round(P_fem_surv, 3))

Подмножество исходов про пассажир — женщина, и одновременно с этим пассажир выжила в катастрофе:
      PassengerId  Survived  Pclass     Sex
1              2         1       1  female
2              3         1       3  female
3              4         1       1  female
8              9         1       3  female
9             10         1       2  female
..           ...       ...     ...     ...
874          875         1       2  female
875          876         1       3  female
879          880         1       1  female
880          881         1       2  female
887          888         1       1  female

[233 rows x 4 columns]

Вероятность такого исхода = 0.262


5. Найдите условную вероятность того, что пассажир выживет, при условии, если пассажир — женщина.


In [13]:
P_cond_surv_fem = P_fem_surv / P_fem
print("Вероятность того, что пассажир выживет, при условии, если пассажир — женщина =", round(P_cond_surv_fem, 3))

Вероятность того, что пассажир выживет, при условии, если пассажир — женщина = 0.742


6. По аналогии найдите вероятность того, что пассажир выживет, если пассажир ехал в первом классе.


In [14]:
X_t6 = X[(X["Survived"] == 1) &  (X["Pclass"] == 1)]
print("Подмножество исходов про 1класс и одновременно выжил:\n", X_t6)

P_surv_1cl = X_t6.shape[0] / X.shape[0]
print("\nВероятность такого исхода =", round(P_surv_1cl, 3))

P_cond_surv_1class = P_surv_1cl / P_1class
print("Вероятность, что пассажир выживет, при условии 1 класса путешествия =", round(P_cond_surv_1class, 3))

Подмножество исходов про 1класс и одновременно выжил:
      PassengerId  Survived  Pclass     Sex
1              2         1       1  female
3              4         1       1  female
11            12         1       1  female
23            24         1       1    male
31            32         1       1  female
..           ...       ...     ...     ...
862          863         1       1  female
871          872         1       1  female
879          880         1       1  female
887          888         1       1  female
889          890         1       1    male

[136 rows x 4 columns]

Вероятность такого исхода = 0.153
Вероятность, что пассажир выживет, при условии 1 класса путешествия = 0.63


7. Найдите вероятность того, что пассажир выживет при условии, что пассажир ехал в первом классе и является женщиной.


In [15]:
X_t7 = X[(X["Pclass"] == 1) & (X["Sex"] == "female")]
P_1cl_fem = X_t7.shape[0] / X.shape[0]
print("Вероятность исхода про 1класс и женщина =", round(P_1cl_fem, 3))

X_t7_surv = X[(X["Survived"] == 1) & (X["Pclass"] == 1) & (X["Sex"] == "female")]
P_surv_1cl_fem = X_t7_surv.shape[0] / X.shape[0]
print("Вероятность исхода про про 1класс и женщина, и выжил =", round(P_surv_1cl_fem, 3))

P_cond_surv_1class_fem = P_surv_1cl_fem / P_1cl_fem
print("\nВероятность, что пассажир выживет, при условии 1 класса и женского пола =", round(P_cond_surv_1class_fem, 3))

Вероятность исхода про 1класс и женщина = 0.105
Вероятность исхода про про 1класс и женщина, и выжил = 0.102

Вероятность, что пассажир выживет, при условии 1 класса и женского пола = 0.968


8. Найдите вероятность того, что пассажир выживет при условии, что пассажир ехал во втором классе и является мужчиной.


In [16]:
X_t8 = X[(X["Pclass"] == 2) & (X["Sex"] == "male")]
P_2cl_male = X_t8.shape[0] / X.shape[0]
print("Вероятность исхода про 2класс и мужчина =", round(P_2cl_male, 3))

X_t8_surv = X[(X["Survived"] == 1) & (X["Pclass"] == 2) & (X["Sex"] == "male")]
P_surv_2cl_male = X_t8_surv.shape[0] / X.shape[0]
print("Вероятность исхода про про 2класс и мужчина, и выжил =", round(P_surv_2cl_male, 3))

P_cond_surv_2class_male = P_surv_2cl_male / P_2cl_male
print("\nВероятность, что пассажир выживет, при условии 2 класса и мужского пола =", round(P_cond_surv_2class_male, 3))

Вероятность исхода про 2класс и мужчина = 0.121
Вероятность исхода про про 2класс и мужчина, и выжил = 0.019

Вероятность, что пассажир выживет, при условии 2 класса и мужского пола = 0.157


9. Используя посчитанные вероятности, найдите вероятность того, что случайно выбранный пассажир — женщина, если мы знаем, что пассажир выжил.


In [17]:
P_cond_fem_surv = P_fem_surv / P_surv
print("\nВероятность, что пассажир - женщина, при условии выживания =", round(P_cond_fem_surv, 3))


Вероятность, что пассажир - женщина, при условии выживания = 0.681


10. Используя посчитанные вероятности, найдите вероятность того, что случайно выбранный пассажир ехал в первом классе, если мы знаем, что пассажир выжил.


In [18]:
P_cond_1class_surv = P_surv_1cl / P_surv
print("\nВероятность, что пассажир из 1кл, при условии выживания =", round(P_cond_1class_surv, 3))


Вероятность, что пассажир из 1кл, при условии выживания = 0.398


11. Сравните полученные вероятности с безусловными вероятностями соответствующих событий. Объясните разницу в значениях.

In [19]:
print("Средний шанс на выживание на рейсе =", round(P_surv,3))
print("Средний шанс на выживание для женщин  =", round(P_cond_surv_fem,3))

print("\nБезусловная вероятность быть выжившей женщиной =", round(P_fem_surv,3))
print("Условная вероятность оказаться женщиной при выживании =", round(P_cond_fem_surv, 3))

Средний шанс на выживание на рейсе = 0.384
Средний шанс на выживание для женщин  = 0.742

Безусловная вероятность быть выжившей женщиной = 0.262
Условная вероятность оказаться женщиной при выживании = 0.681


Женщин выжило непропорционально много (74% вместо 38%), поэтому при условии "пассажир выжил" условная вероятность "это была женщина" получается выше, чем безусловная. Логично  

In [20]:
print("Средний шанс на выживание на рейсе =", round(P_surv,3))
print("Средний шанс на выживание для 1кл  =", round(P_cond_surv_1class,3))

print("\nБезусловная вероятность быть выжившим 1кл =", round(P_surv_1cl,3))
print("Условная вероятность оказаться 1кл при условии выживании =", round(P_cond_1class_surv, 3))

Средний шанс на выживание на рейсе = 0.384
Средний шанс на выживание для 1кл  = 0.63

Безусловная вероятность быть выжившим 1кл = 0.153
Условная вероятность оказаться 1кл при условии выживании = 0.398


Пассажиров 1 класса выжило непропорционально много (63% вместо 38%), поэтому при условии "пассажир выжил" условная вероятность "являться пассажиром 1 класса" оказалась выше, чем безусловная. Логично  

### **Задание 3**
Рассмотрим датасет со статистикой почтового ящика `letters.xlsx`. Данные можно найти в материалах к практической работе. Там содержится 248 записей с информацией о письмах. 

Нам хочется узнать, является ли новое письмо, пришедшее к нам на почту, спамом. Для этого у нас существует статистика писем. Каждое пришедшее письмо может быть либо спамом, либо нет. При этом и спам, и нормальные письма могут содержать слова, характерные для спам-писем. Данный датасет представляет собой набор записей с информацией о пришедших ранее письмах, распознанных человеком как спам или нет. 




In [21]:
df = pd.read_excel('letters.xlsx', sheet_name='Sheet1', index_col='id_letter')
df

spam spam_words
id_letter                
1           да        нет
2           да         да
3          нет        нет
4          нет         да
5           да         да
...        ...        ...
244        нет        нет
245        нет        нет
246        нет        нет
247        нет        нет
248        нет        нет

[248 rows x 2 columns]

Столбец `spam` отвечает на вопрос, является письмо спамом или нет, `spam_words` — содержит ли оно слова, встречающиеся в спам-письмах.

1. Найдите долю спам-писем по отношению ко всем.



In [22]:
spam_data = df["spam"].value_counts()
P_spam = spam_data["да"] / sum(spam_data)

print(spam_data, "\nДоля спам-писем =", round(P_spam, 3))

spam
нет    160
да      88
Name: count, dtype: int64 
Доля спам-писем = 0.355


2. Найдите долю писем со спам-подобными словами по отношению ко всем.


In [23]:
words = df["spam_words"].value_counts()
P_words = words["да"] / sum(words)

print(words, "\nДоля писем со спам-словами =", round(P_words, 3))

spam_words
нет    155
да      93
Name: count, dtype: int64 
Доля писем со спам-словами = 0.375


3. Найдите вероятность, что письмо содержит спам-подобные слова при условии, что оно является спамом.


In [24]:
words_spam = df[(df["spam"] == "да") & (df["spam_words"] == "да")]
P_words_spam = words_spam.shape[0] / df.shape[0]

P_cond_words_spam = P_words_spam / P_spam
print("\nВероятность наличия спам-слов для спама =", round(P_cond_words_spam, 3))


Вероятность наличия спам-слов для спама = 0.364


4. Найдите вероятность, что письмо является спамом при условии, что оно содержит спам-подобные слова.


In [25]:
P_cond_spam_words = P_words_spam / P_words
print("\nВероятность, что письмо является спамом, при условии наличия спам-слов =", round(P_cond_spam_words, 3))


Вероятность, что письмо является спамом, при условии наличия спам-слов = 0.344


### Вывод
Вероятность выбрать спам-письмо взяв случайное письмо = 0.355
<p>
Вероятность выбрать спам-письмо взяв "подозрительное" письмо = 0.344
<p>
Метод маркировки "подозрительных" писем выбран неудачно(